src/span.rs

```diff
@@ -0,0 +1,142 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#[cfg(not(feature = "std"))]
+use alloc::{boxed::Box, vec::Vec};
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+/// A byte span within the parsed string
+#[derive(Debug, Eq, Clone, Copy)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum Span {
```

Comment:
What about making this a wrapper for `Token`? 

Like:
```rust
struct WithSpan {
    token: Token,
    // position
    ...
}
```

Then you don't have to modify parser heavily, but wrapping the error with span at top of parser instead.

Reply:
This is indeed a better approach. Since there did not seem to be much interest in this MR. I have started on my [own parser]( https://github.com/antialize/sql-ast) to experiment with spans and proper error recovery:
Here I ended up having the lexer return `(Token<'a>, Span)`
```rust
pub fn next_token(&mut self) -> (Token<'a>, Span) {
```
I have also created a [Spanned](https://github.com/antialize/sql-ast/blob/b0c2ce277ac08b31c5e2034e479f66a1ff24eb30/src/span.rs#L29) trait that is implemented for all AST nodes.

I need a parser that can produce spans for tokens and nodes, that can report multiple errors, and return an AST even if it encountered some errors.

I tried to see if I could add it to this crate, but it seems hard without breaking all existing uses of the library, perhaps it can be done somehow.